CA Yuhang Liao

ComarPers922 opened the pull request.

lepidopterane-atsmith committed:
CA Tester V1.0.0

This is incomplete but here’s a look at what I’m working on. It’s based
on the BasicControllerTest. Please read my note before testing (though
this is not ready to run yet).

ComarPers922 committed:
Prevent denied users from reading messages.

ComarPers922 committed:
Merged!

lepidopterane-atsmith committed:
CA Tester V1.0.1

Still incomplete, but this commit reflects that I’m 70% there.

lepidopterane-atsmith commented on src/codeu/chat/client/core/MessageContext.java:
```diff
@@ -18,18 +18,31 @@
+=======
 import codeu.chat.util.Serializer;
 import codeu.chat.util.Serializers;
+>>>>>>> e8b7c466b90f32b9119079c9b925bdbe3d30292d
```
FYI, you should delete 25-8. I added those lines back when I had no idea how serialization in this project works.

lepidopterane-atsmith commented on src/codeu/chat/client/core/MessageContext.java:
```diff
@@ -18,18 +18,31 @@
   
   public final Message message;
+<<<<<<< HEAD
+  private final ConversationHeader conversation;
```
MessageContext is just a class that uses a View and a Message to deal with Messages on the client-end (like a lot of classes named ______Context), and in fact, you can use Message's variables to find both its parent conversation and its author, if you look at the Message.java class.

lepidopterane-atsmith commented on src/codeu/chat/client/core/View.java:
```diff
@@ -120,18 +120,28 @@ public View(ConnectionSource source) {
 
   @Override
-  public Collection<Message> getMessages(Collection<Uuid> ids) {
+  public Collection<Message> getMessages(ConversationHeader conversation, User user, Collection<Uuid> ids) {
```
I learned the hard way that it's better to get the Uuid from the user and have that go back and forth between client and server, because really we only need something to identify our users by, no other info in User, and Uuid has more stable serialization than User anyway. Also, User does weird things when you try to put it in a HashMap/HashSet.

lepidopterane-atsmith commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,6 +34,11 @@
+  public static final byte USER_TYPE_CREATOR = 0x07;
+  public static final byte USER_TYPE_OWNER = 0x03;
+  public static final byte USER_TYPE_USER = 0x01;
+  public static final byte USER_TYPE_BANNED = 0x00;
```
I wasn't sure what you were talking about when you brought this idea up, but I have to say: This is a smart way of implementation I didn't consider, so nice job with that.

lepidopterane-atsmith commented on src/codeu/chat/server/Model.java:
```diff
@@ -72,7 +76,11 @@ public int compare(Time a, Time b) {
   private final Store<String, Message> messageByText = new Store<>(STRING_COMPARE);
 
+<<<<<<< HEAD
+  private final LinkedHashMap<Uuid, LinkedHashMap<Uuid, Byte>> authority = new LinkedHashMap<>();
```
While I'm ok with your use of LinkedHashMap, what bothers me is that I thought we agreed that we'd use something like `LinkedHashMap<ConversationHeader, LinkedHashMap<Uuid, Byte>>` so we could just grab something to work with once. Even the doc reflects that, and so do my tests.

lepidopterane-atsmith commented on src/codeu/chat/server/Model.java:
```diff
@@ -134,6 +142,24 @@ public void add(Message message) {
+    LinkedHashMap<Uuid,Byte> userAuthority;
+    if(authority.get(conversation.id) == null)
+    {
+       authority.put(conversation.id, new LinkedHashMap<>());
```
Clever. But use `conversation` not `conversation.id`, really, it'll help in the long run!

lepidopterane-atsmith commented on src/codeu/chat/server/Model.java:
```diff
@@ -134,6 +142,24 @@ public void add(Message message) {
+  {
+    LinkedHashMap<Uuid, Byte> userAuthority = authority.get(conversation.id);
+    return userAuthority.get(conversation.id) != null && userAuthority.get(targetUser).byteValue() == authorityByte;
+=======
```
Where's the } in this method?? For next time, know that if someone forgets a semicolon or one of these>> ] } ) it's a sign that they didn't compile their code before committing. I understand if you're just pushing these incomplete things now, but making this kind of mistake in the code review phase (which we're not in, technically this is a ghost review) would be a big no-no, and by then you'd need to compile anyway.

lepidopterane-atsmith replied:
oopsadaisy I didn't see the matching }

lepidopterane-atsmith reviewed (changes requested):
Some things were clever, more things could use improvement. We don't have to worry about merge conflicts just yet, but when we get there, I'll help you through it.

lepidopterane-atsmith commented:
Actually, can you just tell me how you think this feature works from the moment m-list is called? How does all the code you wrote in different classes flow? What steps happen?

ComarPers922 committed:
Merged!

aslo commented on src/codeu/chat/client/core/ConversationContext.java:
```diff
@@ -87,9 +87,15 @@ private ConversationPayload getUpdated() {
     return payloads.hasNext() ? payloads.next() : null;
   }
 
+<<<<<<< HEAD
```
These are merge conflict markers, and you should not commit code in an unmerged state like this. This is also bad because the code doesn't compile.  Please resolve these merge conflicts.

aslo commented on src/codeu/chat/client/core/MessageContext.java:
```diff
@@ -18,18 +18,31 @@
 import codeu.chat.common.BasicView;
+import codeu.chat.common.ConversationHeader;
 import codeu.chat.common.Message;
+<<<<<<< HEAD
```
See my above comment - this merge conflict needs to be resolved as well.

aslo commented on src/codeu/chat/client/core/MessageContext.java:
```diff
@@ -18,18 +18,31 @@
+  private final BasicView view;
+=======
   public final BasicView view;
+>>>>>>> e8b7c466b90f32b9119079c9b925bdbe3d30292d
```
See my above comment - this merge conflict needs to be resolved as well.

aslo commented on src/codeu/chat/common/BasicView.java:
```diff
@@ -51,7 +51,11 @@
   // GET MESSAGES
   //
   //   Return all messages whose id is found in the given collection.
+<<<<<<< HEAD
```
See my above comment - this merge conflict needs to be resolved as well.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,6 +34,11 @@
 
 public final class Controller implements RawController, BasicController {
 
+  public static final byte USER_TYPE_CREATOR = 0x07;
```
Should this be `0x04`? You only need a single bit to represent each role, and it's possible for a user to be a creator without being an owner or a user, right? If you represent the creator, owner, and user roles as `0b100`, `0b010`, and `0b001` respectively, your code can combine those in any possible combination.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,6 +34,11 @@
 public final class Controller implements RawController, BasicController {
 
+  public static final byte USER_TYPE_CREATOR = 0x07;
+  public static final byte USER_TYPE_OWNER = 0x03;
```
See above comment - I think this should be `0x02`

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -15,7 +15,11 @@
+import java.util.LinkedHashMap;
+=======
 import java.util.HashMap;
+>>>>>>> e8b7c466b90f32b9119079c9b925bdbe3d30292d
```
Same as above - remove these merge artifacts.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -134,6 +142,24 @@ public void add(Message message) {
   }
+<<<<<<< HEAD
+
+  public void changeAuthority(ConversationHeader conversation, Uuid targetUser, byte authorityByte)
```
Don't forget - you should validate that the user has the authority to call this method! I don't believe you are currently checking it.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -134,6 +142,24 @@ public void add(Message message) {
+  public boolean validateAuthority(ConversationHeader conversation, Uuid targetUser, byte authorityByte)
+  {
+    LinkedHashMap<Uuid, Byte> userAuthority = authority.get(conversation.id);
+    return userAuthority.get(conversation.id) != null && userAuthority.get(targetUser).byteValue() == authorityByte;
```
This isn't quite right - remember that the user doesn't need to exactly match the required `authorityByte` - if they have a superset of the required authority bits, they should still be allowed to perform the action in question. (If the user has bits `011`, but the `authorityByte` argument is `001`, this method should still return true.

You can use [bitwise operations](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/op3.html) to do this in a clever way!

aslo reviewed (changes requested):
Looking good - I left a few comments, but the biggest issue is that you committed a lot of merge markers with your code, so the code in this state doesn't compile. Don't forget - always make sure your code at the very least can compile before submitting PRs.

ComarPers922 committed:
Implement the major part of the feature.

lepidopterane-atsmith committed:
CA Tester V1.0.2

Here it is, the first finished version of the comprehensive JUnit
Tester for CA. Please let me know if you have any
questions/concerns/whatever.

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -417,6 +418,8 @@ public void invoke(List<String> args) {
         System.out.println("    List all messages in the current conversation.");
         System.out.println("  m-add <message>");
         System.out.println("    Add a new message to the current conversation as the current user.");
+        System.out.println("  u-authorize <user> <authority>");
```
Hey, I'm just gonna leave some suggestions here to help you, so once again, this is a premature code review that doesn't fully count. Whenever you're on one menu of commands in this app, they all start with the same letter: u-menu is when you boot up, c-menu once you login, m-menu for conversation commands, s-menu for status update. So this should be m-authorize just to keep everything organized. Also, please explain authority markers in this dialog.

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -425,7 +428,34 @@ public void invoke(List<String> args) {
+      {
+        if (args.size() < 2) 
+        {
+      	  System.out.println("ERROR: Please provide enough parameters!!!");
```
It would be more helpful if you provided a message reminding the user of the command format, such as,
"ERROR: Command does not follow the format m-authorize <user> <authority>"

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -425,7 +428,34 @@ public void invoke(List<String> args) {
+          else if(!args.get(1).equals("o") && !args.get(1).equals("u") && !args.get(1).equals("b"))
+          {
+            System.out.println(args.get(1));
+            System.out.println("ERROR: Parameter is unacceptable!");
```
Once again, it would be more helpful if the string looked something like this:
"ERROR: Authority "+args.get(1)+" not supported. Use help to learn more."

lepidopterane-atsmith commented on src/codeu/chat/server/Model.java:
```diff
@@ -148,11 +149,75 @@ public void changeAuthority(ConversationHeader conversation, Uuid targetUser, by
+  }
+
+
+  public boolean isUser(Uuid conversation,Uuid targetUser)
```
You should change both of your isUser methods to isMember because otherwise we don't know what you mean. A good way to think about it is that just like squares are a sort of rectangle, members are a kind of user.

lepidopterane-atsmith commented on src/codeu/chat/server/Model.java:
```diff
@@ -148,11 +149,75 @@ public void changeAuthority(ConversationHeader conversation, Uuid targetUser, by
+  public boolean isBannedUser(ConversationHeader conversation,Uuid targetUser)
+  {
+    LinkedHashMap<Uuid, Byte> userAuthority = authority.get(conversation.id);
+    Byte authoritByte = userAuthority.get(targetUser);
```
Stupid question: If you made this a `byte` instead of a `Byte`, would the code still work? I think it does but I'm not sure...

lepidopterane-atsmith commented on src/codeu/chat/server/Server.java:
```diff
@@ -81,7 +81,30 @@ public Server(final Uuid id, final Secret secret, final Relay relay,final File l
     this.controller = new Controller(id, model,localFile);//Use the new constructor to create this new controller.
     this.relay = relay;
-
+    this.commands.put(NetworkCode.CONVERSATION_AUTHORITY_REQUEST, new Command()
```
Stick this added chunk at the bottom of the this.commands.put(NetworkCode... section. If you notice, they're all grouped by Network Code number, so putting a big number in front of a small number like this is a little awkward.

lepidopterane-atsmith commented on src/codeu/chat/server/Server.java:
```diff
@@ -81,7 +81,30 @@ public Server(final Uuid id, final Secret secret, final Relay relay,final File l
+        {
+          Serializers.INTEGER.write(out, NetworkCode.CONVERSATION_ACCESS_DENIED);
+        }
+        else if(model.isOwner(conversation, fromUser.id) && model.isCreator(conversation, targetUser.id))
```
Personally, you could get a lot more done if you merge the else if and if into 
`if( model.isUser(conversation, fromUser.id) || model.isBannedUser(conversation, fromUser.id) || model.isCreator(conversation, targetUser.id))` because not even a Creator should be able to accidentally demote themselves.

lepidopterane-atsmith reviewed (commented):
Just some suggestions to help you along...

lepidopterane-atsmith committed:
CA Tester V1.0.3

Whoops, forgot to actually plug this into TestRunner!

lepidopterane-atsmith committed:
CA Test Suite V1.0.4

Forgot some things in ConversationAccessServerTest (CAST). Also
demonstrates what I’ve been working on lately.

lepidopterane-atsmith committed:
CA Test Suite V1.0.5

Trying to ungank this project so it doesn’t confuse Eclipse.

lepidopterane-atsmith committed:
CA Test Suite V1.0.6

Still polishing some odds and ends…

lepidopterane-atsmith committed:
CA Test Suite V1.0.7

There, that’s better.

lepidopterane-atsmith committed:
CA Test Suite 1.0.8

Further cleaning.

lepidopterane-atsmith committed:
CA Test Suite 1.0.9

Now tests the different nonpersistent storage mechanism I’ve made for
this feature.

ComarPers922 committed:
1. Implement the new interface.  2. Delete all NotImplementedExceptions 3.Use HashMap 4. Change the byte values for users in codeu.chat.server.Controller

lepidopterane-atsmith committed:
CA Test Suite V1.1.0

Polished some things. Refactored testAccessSecurity() into 16 separate
tests.

lepidopterane-atsmith committed:
CA Test Suite V1.1.0.5

Further polishing...

lepidopterane-atsmith committed:
CA Test Suite V1.1.1

No more tests containing 200 lines and 20 asserts! These are refactored
now.

aslo commented on src/codeu/chat/server/AuthorityModel.java:
```diff
@@ -0,0 +1,13 @@
+public interface AuthorityModel
```
This interface has compile errors because a few things are missing: imports, and the package declaration. Can you please add them?

aslo reviewed (changes requested):
This is looking great! Just one small thing to fix and then I'm happy to approve.

ComarPers922 committed:
Please implement the interface!

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -34,6 +34,7 @@
 import codeu.chat.common.InterestSet;
 import codeu.chat.common.ServerInfo;
 import codeu.chat.common.User;
+import codeu.chat.server.Controller;
```
You should delete this since 1) I don't see it used anywhere and 2) if you're working in a client file, never import a server class and vice versa, everything just gets really disordered and confusing

lepidopterane-atsmith commented on src/codeu/chat/client/core/ConversationContext.java:
```diff
@@ -56,7 +56,7 @@ public MessageContext add(String messageBody) {
     return message == null ?
         null :
-        new MessageContext(message, view);
+        new MessageContext(message, conversation, user, view);
```
Once again, you don't really need to add these new variables, both conversation and user the message belongs to are Uuid variables in the Message class. I understand you need both variables as non-Uuid parameters for authorityModificationRequest rn, but it's better to take them as Uuids in the method that you can grab a variable from later.

lepidopterane-atsmith commented on src/codeu/chat/client/core/ConversationContext.java:
```diff
@@ -102,4 +102,8 @@ public MessageContext findMessageByUuid(Uuid id) {
     return null;
   }
   
+  public void changeAuthority(User targetUser, String authorityParameter)
```
Oh yeah, and targetUser should be a Uuid! If all these Uuids are confusing you can write nested classes that will clear things up nicely.

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -425,7 +428,34 @@ public void invoke(List<String> args) {
+          {
+            System.out.println("ERROR: No such user.");
+          }
+          else if(!para.equals("o") && !para.equals("u") && !para.equals("b"))
```
Also, because members are a kind of user, can we please change the "u" rank to an "m" rank? It just makes more sense.

lepidopterane-atsmith commented on src/codeu/chat/client/core/MessageContext.java:
```diff
@@ -18,18 +18,22 @@
+  private final BasicView view;
 
-  public MessageContext(Message message, BasicView view) {
+  public MessageContext(Message message, ConversationHeader conversation, User user, BasicView view) {
```
I'll say it once and I'll say it again: any variable you need to get conversation and user is in the Message class itself.

lepidopterane-atsmith commented on src/codeu/chat/server/Controller.java:
```diff
@@ -33,6 +33,11 @@
+  public static final byte USER_TYPE_OWNER = 0x03;
+  public static final byte USER_TYPE_USER = 0x01;
+  public static final byte USER_TYPE_BANNED = 0x00;
 
```
`creatorByte = 0b111;`
`ownerByte = 0b011;`
`memberByte (member not user!) = 0b001;`
`bannedByte = 0b000;`

lepidopterane-atsmith reviewed (changes requested):
Took another look at this while modifying my test suite, overall looks pretty good but there's still some things that could be refactored to make life easier.

ComarPers922 committed:
Please check this new version.

ComarPers922 committed:
Fewer methods in AuthorityModel should be implemented!

lepidopterane-atsmith committed:
CA Test Suite V1.1.2

Addressed Alex’s most recent comments and also incorporated more of the
actual commands into testing.

lepidopterane-atsmith committed:
Model now implements AuthorityModel!

Glad that’s done!

lepidopterane-atsmith committed:
CA Test Suite now operational

Successfully grafted all of Yuhang’s code into my environment. Now we
can put this tester to good use!

lepidopterane-atsmith committed:
CA TestSuite V1.1.3

Fixed some errors under my responsibility.

ComarPers922 committed:
Good to go!

ComarPers922 committed:
All finished!

ComarPers922 committed:
Almost full implementation of CA with persistent storage. Delete extra imports.

lepidopterane-atsmith committed:
Help me non-null this please

Hey, this is Sarah, not Yuhang. I had a weird github issue dealing with
the code in addition to this null issue so sorry I’m pushing here
Yuhang but please help me work this out!!!!!!!
